libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs

```diff
@@ -0,0 +1,222 @@
+        /// <param name="dialogId">ID of the prompt to start.</param>
+        /// <param name="prompt">Initial prompt to send the user.</param>
+        /// <param name="choicesOrOptions">(Optional) array of choices to prompt the user for or additional prompt options.</param>
+            
+        public Task Prompt(string dialogId, string prompt, PromptOptions options = null)
+        {
+            if (options == null)
+            {
+                options = new PromptOptions();
+            }
+            if (prompt != null)
+            {
+                options.PromptString = prompt;
+            }
+            return Begin(dialogId, options);
+        }
+        public Task Prompt(string dialogId, Activity prompt, PromptOptions options = null)
+        {
+            if (options == null)
+            {
+                options = new PromptOptions();
+            }
+            if (prompt != null)
+            {
+                options.PromptActivity = prompt;
+            }
+            return Begin(dialogId, options);
+        }
+
+        /// <summary>
+        /// Continues execution of the active dialog, if there is one, by passing the context object to
+        /// its `Dialog.continue()` method. You can check `context.responded` after the call completes
+        /// to determine if a dialog was run and a reply was sent to the user.
+        /// </summary>
+        public async Task Continue()
+        {
+            // Check for a dialog on the stack
+            if (Instance != null)
+            {
+                // Lookup dialog
+                var dialog = Dialogs.Find(Instance.Id);
+                if (dialog == null)
+                {
+                    throw new Exception($"DialogSet.continue(): Can't continue dialog. A dialog with an id of '{Instance.Id}' wasn't found.");
+                }
+
+                // Check for existence of a continue() method
+                if (dialog.HasDialogContinue)
+                {
+                        // Continue execution of dialog
+                        await dialog.DialogContinue(this);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Ends a dialog by popping it off the stack and returns an optional result to the dialogs
+        /// parent.The parent dialog is the dialog the started the on being ended via a call to 
+        /// either[begin()](#begin) or [prompt()](#prompt). 
+
+        /// The parent dialog will have its `Dialog.resume()` method invoked with any returned
+        /// result. If the parent dialog hasn't implemented a `resume()` method then it will be
+        /// automatically ended as well and the result passed to its parent. If there are no more
+        /// parent dialogs on the stack then processing of the turn will end.
+        /// </summary>
+        /// @param result (Optional) result to pass to the parent dialogs `Dialog.resume()` method.
+        public async Task End(object result = null)
```

Comment:
End == Pop 

WHAT?? It's either a "Dialog Stack" or it's not.

Reply:
In reading the code below, the recursion is unexpected. 

The semantic meaning of "End" is NOT "Pop". It's "Pop, plus mutate and do stuff to something else". 

That behaivor is actually reasonable, given my understanding of the Dialog constructs here, but the NAME of "End" is seems misleasing. "EndCurentAndResultPrevious" or something would be better.

Reply:
Again... this is what its called in v3.  If I was to rename it I'd probably call it `return()` but `end()` is fine and more importantly familiar if you're a Node developer.

All of this is going to be foreign to C# developers but at least we will have parity now.